src/main/java/de/tum/in/www1/artemis/service/AchievementService.java

```diff
@@ -0,0 +1,118 @@
+            TimeBasedAchievementService timeBasedAchievementService, ProgressBasedAchievementService progressBasedAchievementService) {
+        this.achievementRepository = achievementRepository;
+        this.pointBasedAchievementService = pointBasedAchievementService;
+        this.timeBasedAchievementService = timeBasedAchievementService;
+        this.progressBasedAchievementService = progressBasedAchievementService;
+        this.userRepository = userRepository;
+    }
+
+    /**
+     * Finds all achievements for a user in a given course and returns them as a set
+     * @param userId
+     * @param courseId
+     * @return set of achievements
+     */
+    public Set<Achievement> findAllByUserIdAndCourseId(Long userId, Long courseId) {
+        return achievementRepository.findAllByUserIdAndCourseId(userId, courseId);
+    }
+
+    /**
+     * Deletes all achievements that belong to the course with the given courseId
+     * Used when a course is deleted or when achievements are disabled again for a course
+     * @param courseId
+     */
+    public void deleteByCourseId(Long courseId) {
+        Set<Achievement> achievements = achievementRepository.findAllByCourseId(courseId);
+        for (Achievement achievement : achievements) {
+            removeFromUsers(achievement);
+        }
+        achievementRepository.deleteAll(achievements);
+    }
+
+    /**
+     * Deletes all achievements that belong to the exercise with the given exerciseId
+     * Used when an exercise is deleted
+     * @param exerciseId
+     */
+    public void deleteByExerciseId(Long exerciseId) {
+        Set<Achievement> achievements = achievementRepository.findAllByExerciseId(exerciseId);
+        for (Achievement achievement : achievements) {
+            removeFromUsers(achievement);
+        }
+        achievementRepository.deleteAll(achievements);
+    }
+
+    /**
+     * Removes an achievement from all users
+     * @param achievement achievement to be deleted
+     */
+    public void removeFromUsers(Achievement achievement) {
+        var users = achievement.getUsers();
+        for (User user : users) {
+            user.removeAchievement(achievement);
+        }
+        userRepository.saveAll(users);
+    }
+
+    /**
+     * Generates achievements for a course
+     * Used when course is updated or created and achievements are enabled for course
+     * @param course
+     */
+    public void generateForCourse(Course course) {
+        progressBasedAchievementService.generateAchievements(course);
+    }
+
+    /**
+     * Generates achievements for an exercise
+     * If achievements are enabled for corresponding course and exercise is not part of exam
+     * Used when exercise is created
+     * @param course
+     * @param exercise
+     */
+    public void generateForExercise(Course course, Exercise exercise) {
+        if (course.getHasAchievements() && (exercise.getExerciseGroup() == null || exercise.getExerciseGroup().getExam() == null)) {
```

Comment:
You could use `exercise.hasExerciseGroup()` to check for exam exercises.

Reply:
If that is sufficient, sure.